IIIF REST API

mspalti opened the pull request:
## References

* Related to Angular PR: https://github.com/DSpace/dspace-angular/pull/1079 See that issue for an overview.

## Description
This pull request implements IIIF endpoints for manifests and annotation lists (Presentation API 2.1.1) and searches (Search API 3.0).  It provides an optional solr core for searching within documents and adds dspace configuration properties for search and image services.

## Instructions for Reviewers (updated on 2021.09.30)

List of changes in this PR:
* Added new dependency on IIIF model: https://github.com/dbmdz/iiif-apis
* Added REST controller, repository, and services for IIIF.
* Added model classes.
* Added DSpace IIIF configuration options.

**Include guidance for how to test or review your PR.** This may include: steps to reproduce a bug, screenshots or description of a new feature, or reasons behind specific changes.

**DSpace Items** that will be exposed via the IIIF endpoint **require** to have the `dspace.iiif.enabled` metadata set to `true`. 
The functionality makes use of additional metadata at the item and bitstream level to override settings otherwise defined in the configuration. For a list of metadata and their purpose look to the `iiif-types.xml` metadata file registry introduced by this PR.

**DSpace Configuration**: See examples provided in `iiif.cfg`. Defaults are ok **for API testing**, but you will need to provide an image server if you want to see results in the Angular UI. The easier approach is to use cantaloupe and configure it to use an http source, see https://cantaloupe-project.github.io/manual/5.0/sources.html#HttpSource
the configuration will be
```
HttpSource.BasicLookupStrategy.url_prefix = <dspace-url>/server/api/core/bitstreams/
HttpSource.BasicLookupStrategy.url_suffix = /content
```
 
### IIIF REST Endpoint Samples 

**Sample manifest:** 
[https://digitalcollections.willamette.edu/server/iiif/0a53d851-319b-4883-9d80-6bf0148ee2ed/manifest](https://digitalcollections.willamette.edu/server/iiif/0a53d851-319b-4883-9d80-6bf0148ee2ed/manifest)

Manifests may include any of the following properties: `metadata`, `service` (search), `seeAlso,` `sequence,` `structures`, and `related` as defined in [https://iiif.io/api/presentation/2.1/](https://iiif.io/api/presentation/2.1/). You should be able to see a manifest for any item that meets the requirements above.

Manifests are offered to users that have READ permission over the item, as the manifest itself only contains minimal metadata about the bitstream in the items these are always included, the actual images will be accessible or not according to the permission granted to the IIIF Image Server. By default the IIIF Image Server will be treated as an Anonymous user but you can increase its permissions 
- configuring in DSpace an IP Authentication that will put it in a special group
- implementing a more sophisticated / custom lookup strategy in cantaloupe that will perform some kind of authentication.

In future we could explore support for a tight integration via the IIIF Authentication API.

**Sample search:**

(Experimental) Support for the Search API is disabled by default. There is currently a single search plugin: `org.dspace.app.rest.iiif.service.WordHighlightSolrSearch` that can be configured in `iiif.cfg`.  It depends on a solr core and library described here: [https://github.com/dbmdz/solr-ocrhighlighting](https://github.com/dbmdz/solr-ocrhighlighting).  The library jar must be downloaded from [https://github.com/dbmdz/solr-ocrhighlighting/releases](https://github.com/dbmdz/solr-ocrhighlighting/releases) and added to your solr instance.

**Sample IIIF seeAlso annotation list:**
[https://digitalcollections.willamette.edu/server/iiif/0a53d851-319b-4883-9d80-6bf0148ee2ed/manifest/seeAlso](https://digitalcollections.willamette.edu/server/iiif/0a53d851-319b-4883-9d80-6bf0148ee2ed/manifest/seeAlso)

### Overview of Functionality and Requirements

### IIIF resources

On the Angular side, the metadata `dspace.iiif.enabled` trigger the embed of the Mirador viewer component. On the REST side, the `iiif.search.enabled' metadata prompts the addition of the search service in the manifest. When the search service is included in the manifest, the Mirador viewer will add a search option to the UI.

```
service: {
    @context: "http://iiif.io/api/search/0/context.json",
    @id: "http://localhost:8080/server/api/iiif/85fa7ea8-6ef1-4eec-ae04-9ca587a12767/manifest/search",
    profile: "http://iiif.io/api/search/0/search"
}
```
The search service depends on Solr.  See the Search API discussion below.

### Bundles 

DSpace bundles are important in this implementation.  Special DSpace Bundles other than the ORIGINAL one (license, text, thumbnail, etc) are automatically ignored in the generation of the manifest. All the other Item Bundle, including the ORIGINAL one will be considered. When multiple BUNDLE are available the Bundle is used to generate a Range in the iiif manifest (i.e. a section inside the Table of Contents), in basic scenario only a single ORIGINAL bundle will be considered and no additional range will be created for him. Other bundle can be flagged to be ignored using the `iiif.enabled` metadata.

Images located in the `IIIF` bundles are processed by the REST manifest service and added directly to the canvas sequence.  Here's a example of a `canvas` that has been generated for a bitstream in the IIIF bundle.

**Single canvas example**
```
{
  @id: "http://localhost:8080/server/api/iiif/85fa7ea8-6ef1-4eec-ae04-9ca587a12767/canvas/c0",
  @type: "sc:Canvas",
  label: "Cover",
  thumbnail: {
     @id: "http://localhost:8182/iiif/2/b2e7d92f-0272-4f2c-ad64-917398e39258/full/90,/0/default.jpg",
     service: {
       @context: "http://iiif.io/api/image/2/context.json",
       @id: "http://localhost:8182/iiif/2/b2e7d92f-0272-4f2c-ad64-917398e39258",
       profile: "http://iiif.io/api/image/2/level0.json",
       protocol: "http://iiif.io/api/image"
     },
    format: "image/jp2"
  }, 
  images: [
    {
        @type: "oa:Annotation",
        motivation: "sc:painting",
        resource: {
            @id: "http://localhost:8182/iiif/2/b2e7d92f-0272-4f2c-ad64-917398e39258/full/full/0/default.jpg",
            @type: "dctypes:Image",
            service: {
                @context: "http://iiif.io/api/image/2/context.json",
                @id: "http://localhost:8182/iiif/2/b2e7d92f-0272-4f2c-ad64-917398e39258",
                profile: "http://iiif.io/api/image/2/level1.json",
                protocol: "http://iiif.io/api/image"
            },
            format: "image/jp2"
        },
        on: "http://localhost:8080/server/api/iiif/85fa7ea8-6ef1-4eec-ae04-9ca587a12767/canvas/c0"
    }
  ],
  width: 2613,
  height: 3409
}
```

PDF files in the IIIF bundles _are assumed_ to be alternative rendering of the document.  A link is added in the manifest `rendering` property. The label is derived from the bitstream metadata

**rendering**
```
rendering: {
    @id: "http://localhost:8080/server/api/core/bitstreams/3d94e022-6967-434a-88b7-5212f005a656/content",
    label: "Download as PDF",
    format: "application/pdf"
}
```

The implementation also supports a second, optional `OtherContent` bundle that is used in the manifest `seeAlso` property.  Examples of bitstreams that can be added to this bundle are full text transcriptions, ocr files (e.g. ALTO), and METS. These items are not added directly to the manifest but are linked from the manifest `seeAlso` property.

**seeAlso**
```
seeAlso: {
    @id: "http://localhost:8080/server/api/iiif/85fa7ea8-6ef1-4eec-ae04-9ca587a12767/manifest/seeAlso",
    @type: "sc:AnnotationList",
    label: "More descriptions of this resource"
},
```

### Bitstreams

The IIIF `canvas` is a virtual container that represents a page or view and has content resources associated with it. In this implementation canvases are generated for DSpace bitstreams found in any valid IIIF bundle.  

Each `canvas` must have a label to display, and a height and a width as integers.  The bitstream dso can explicit these information via the new iiif metadata, if such information are not available at the bitstream level the implementation will look for them at the item level and finally default to value in the `iiif.cfg` configuration file.

### info.json (not longer used/present)

**UPDATE: The info.json file has (happily) been replaced.  Labels, height, width, ranges, etc. are defined via DSO metadata.**

### Search API and Word Highlighting (Solr)

The search implementation uses the Solr OCR Highlighting plugin at:  [https://dbmdz.github.io/solr-ocrhighlighting/](https://dbmdz.github.io/solr-ocrhighlighting/). From the documentation: "This Solr plugin lets you put word-level OCR text into one or more of you documents’ fields and then allows you to obtain structured highlighting data with the text and its position on the page at query time." You will need to add the plugin `.jar` and a Solr core the current DSpace Solr server (or other server).  The plugin is available here: [https://github.com/dbmdz/solr-ocrhighlighting/releases](https://github.com/dbmdz/solr-ocrhighlighting/releases).  Directions for how to configure the Solr core here: [https://dbmdz.github.io/solr-ocrhighlighting/installation/](https://dbmdz.github.io/solr-ocrhighlighting/installation/).  I can provide a pre-configured core.

### Image Server (external)

The image server location is defined in dspace.cfg. I am using [https://cantaloupe-project.github.io/](https://cantaloupe-project.github.io/).  _Cantaloupe_ can be configured to use DSpace bitstreams as the http image source.  Any IIIF compliant image service can be used provided it can be configured to retrieve from DSpace as the image source.

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

mspalti committed:
Adding iiif-apis dependency

mspalti committed:
Adding iiif-apis dependency

mspalti committed:
Added iiif model generator classes.

mspalti committed:
Added iiif model generator classes.

mspalti committed:
Added factory for objectmapper.

mspalti committed:
Added factory for objectmapper.

mspalti committed:
Added classes for parsing the info.json object.

mspalti committed:
Added classes for parsing the info.json object.

mspalti committed:
Added iiif service and utility classes.

mspalti committed:
Added iiif service and utility classes.

mspalti committed:
Added iiif repository.

mspalti committed:
Added iiif repository.

mspalti committed:
Added iiif controller.

mspalti committed:
Added iiif controller.

mspalti committed:
Added integration tests.

mspalti committed:
Added integration tests.

mspalti committed:
Added new method to the test BitstreamBuilder and moved iiif test package.

mspalti committed:
Added new method to the test BitstreamBuilder and moved iiif test package.

mspalti committed:
Added sample iiif config to dspace.cfg

mspalti committed:
Added sample iiif config to dspace.cfg

mspalti committed:
Minor refactor manifest generator

mspalti committed:
Minor refactor manifest generator

mspalti committed:
Updated the image service thumbnail path.

mspalti committed:
Updated the image service thumbnail path.

mspalti committed:
Updated dspace.cfg comment.

mspalti committed:
Updated dspace.cfg comment.

mspalti committed:
Added exception to search service query.

mspalti committed:
Added exception to search service query.

mspalti committed:
Updated search service to use solr client libraries.

mspalti committed:
Updated search service to use solr client libraries.

mspalti committed:
Added word_highlighing solr core.

mspalti committed:
Updated dspace.cfg to support iiif ITs

mspalti committed:
Minor updates to iiif services.

mspalti committed:
Merge branch 'iiif-endpoint' of https://github.com/mspalti/DSpace into iiif-endpoint

mspalti committed:
Minor updates.

mspalti committed:
Removed unused import.

tdonohue commented:
@mspalti : A quick note on this PR.It looks really massive because it looks like it includes a lot of unrelated changes.  It looks (to me) like your `iiif-endpoint` branch is not up to date on the latest `main`.  So, it's trying to add back into `main` a ton of old Solr stuff which we removed.  

You may want to rebase your `iiif-endpoint` branch on `main`, as that will make this a lot easier to review & it may also let the code coverage check run (which helps us verify if you have sufficient Integration Tests to prove the new code is working as expected).

mspalti commented:
Thanks for the quick response @tdonohue.  I had an unused import in one of my classes and was holding off on the push until I checked other things.  Test are now completing.

I believe the branch is up to date.  Will check to be sure.  However, I think the old solr code that you are seeing is actually a new core that I added (word_highlighting).  It isn't really necessary for this PR although a new core may need to be included eventually (somewhere). I added it here since I thought it might be helpful to reviewers running the code and testing the search api.

I knew I needed a unit test for the search endpoint since that can't be tested with an IT.  It looks like I may also owe you a few extra ITs based on the code coverage report.

mspalti committed:
Removed word_highlighting solr core.

tdonohue commented:
@mspalti : Just a quick note.  As of a moment ago, I've merged #3183, which renames the "relationship.type" metadata field to be "dspace.entity.type".  It looks like (from the description of this PR) you may be using "relationship.type" in this PR...if so, you'll need to update to use "dspace.entity.type" instead.

mspalti committed:
Removed uncessary encoding of solr query.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Updated IT for revised ItemBuilder.

mspalti committed:
Using dspace.entity.type for iiif bundle lookup.

mspalti committed:
Added caching for manifests.

mspalti committed:
Corrected dspace.entity.type metadata field name in utils.

mspalti commented:
Up to date with the new `dspace.entity.type`.

Also, one thing I need to bring up for discussion is the Solr index for IIIF Search.  That index uses
[this Solr plugin](https://dbmdz.github.io/solr-ocrhighlighting/). It seems quite effective.  However writing code to maintain the Solr index is _currently outside the scope_ of this project. It might be challenging to do since the Solr index requires that ocr files reside on the local file system. If the Solr is not running on the same machine as DSpace that becomes more challenging.

I haven't had time to dig into this yet.  It's a problem that needs to be solved if the Search API is to be fully supported by DSpace. I could imagine a process running on the Solr server that communicates with DSpace and syncs or rebuilds the index as required.  However, for now I'm just manually populating the index with a python script.

The Search API index is the only significant maintenance issue (that I can see).  Configuring an external Image server is easy.

mspalti commented:
One more note on the project scope.

I haven't included collections.  I see no barriers to adding collections now if that's deemed necessary. It's just not been a priority and I have not explored it.

mspalti committed:
Corrected multi-word solr queries.

mspalti committed:
Removed unused import.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Updated logger.

mspalti committed:
Trial implementation of manifest caching.

Fixed style errors.

mspalti commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/Application.java:
```diff
@@ -55,6 +56,7 @@
  */
 @SpringBootApplication
 @EnableScheduling
+@EnableCaching
```
I initially enabled a cache for manifests using the spring boot starter cache and ehcache 3. This is mostly a test case for my own benefit.  My sense is that a better strategy is either (1) enable the cache using ehcache2 in a separate config file located in `iiif/cache` or (2) use the current dspace strategy for configuring caches.  Meanwhile, for the spring boot starter cache to work I needed to add the annotation to the application as I have done here.

In general, server-side caching for manifests seems to be reasonable to me and I'm curious if others agree.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Changed incorrect inheritance.

mspalti committed:
Correction to annotationlist generator.

mspalti commented:
On the question of Solr indexing for `IIIFSearchable` entity.types, I'd like to propose that METS and ALTO files be added to the `OtherContent` bundle.  This allows METS/ALTO retrieval via the IIIF Manifest `seeAlto `AnnotationList. I think that basing ALTO indexing on the contents of the METS file also has advantages. I've written a sample application in Go that runs on the same machine as Solr.  The idea is that this application will be called by a controlling process that monitors changes in DSpace collections.  https://github.com/mspalti/alto_indexer

mspalti committed:
Added trackPages parameter to solr query that was introduced in solr-ocrhighlighing v 0.6.0.

This version was released 5 days ago and is not yet available via Maven.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Updated cache config and renamed classes.

mspalti committed:
Updated cache path.

mspalti committed:
Updated info models.

mspalti committed:
Added IIIF bundle to solr stats.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Added 3rd party licenses.

mspalti committed:
Added missing license header.

mspalti committed:
Removed unused import.

mspalti committed:
Adding IIIF cors configuration.

mspalti committed:
Updated java docs.

mspalti committed:
Adding configurable log to manifest.

mspalti committed:
Added logo url to iiif dspace configuration.

mspalti committed:
Updated ITs to work with the new iiif endpoint.

mspalti committed:
Moved the iiif cors configuration to separate iiif.cfg config file.

mwoodiupui commented:
I like the metadata approach.  I've done something similar in a couple of local projects.  IMHO we don't make enough use of the work that was done to extend metadata support to all DSO-derived types.

As for "not combining native DSpace and IIIF field information in the object, if that's a concern," I don't see it as a concern.  DSpace is designed to hold arbitrary metadata.  What we need to do is to find or create an appropriate namespace for IIIF metadata and use that, not just stuff these data into nonstandard or unsuitable "DC" fields.

abollini commented:
hi @mspalti as anticipated in the dev meeting we (4Science) like to be involved in the review and finalization of this contribution. We hope to be able to bring part of the IIIF solution that we built for version 5+ and that is successful used by several institutions.
We plan to contribute directly code changes where needed so there is no rush for you to make adaption of the existing PR to respond to our feedback, of course we like to have a conversation about all the aspect with you and anyone else in the community interested in following this thread.

The first point that we like to address is how the IIIF support is enabled at the item and bitstream level. Our current solution and recommendation is to use the metadata for all support for that.
We cannot rely on the entity concept here because different entity types could benefit of the IIIF support, book, dataset, maps could need their own entity type to proper manage relationship with other object and forcing the use of a IIIF or IIIFSearchable entity type would prevent that.
It would be better to use a metadata at the item and/or bundle and/or bitstream level to flag which content should be included in a manifest removing the constraint around the IIIF bundle. Or, maybe, we can use the bitstream format to decide if it is suitable for IIIF or not. If we remove the IIIF bundle constraint indeed people will be able to enjoy IIIF just creating an item from the UI and uploading some images in it without requiring them to import the object via an AIP.

For these extra metadata we should use the `dspace` schema and/or a `iiif` schema. I would suggest to use `dspace` for anything that would imply the activation of a feature and iiif for metadata (descriptive or technical) that are needed to improve the generated manifest. For instance we can use a `iiif.toc` at the bitstream level to set the label for the canvas, using the special syntax Level0::Level1::Level2 we can also manage an hierarchical toc so that in the IIIF viewer you will see something like
- chapter 1
-- paragraph 1.1
-- paragraph 1.2
--- extra note 1.2.1
- chapter 2
-- paragraph 2.1

etc.

It is also important to make as easy as possible to enable the feature without having the users to immediately face with all the complexity to also support the IIIF search api. We can think to eventually introduce the IIIF Search API at a later stage, the manifest could decide if include or not a link to a search service depending on a configuration so that in your case (or for the most adventurous that are likely to investigate also on the solr annotation server) the link will be present but for normal dspace installation it would be not including without leaving the user with the impression of a broken feature.

mspalti commented:
Hi @abollini, I agree on most points. When working on this PR I tried isolate as much as possible from existing dspace code and metadata and make the fewest possible code changes. So I do think integration along the lines you suggest is needed and I'm happy to hear that the 4science solution has done this. I agree that the `entity.type` is problematic for the reasons you mention. I was also hoping that we could use metadata (at the item and bitstream levels) to generate manifests. I'm glad that this approach seems to be winning the day. And I like the strategy for a hierarchical toc.

For larger documents I still suspect that there's an advantage to the separate IIIF and OtherContent bundles that I propose, simply because of the large number of files involved (e.g. a 200 page document with 200 ocr files). The separate bundle makes it somewhat easier to update metadata and reorder files.  So I'm curious how you addressed this issue in the 4Science solution. I agree that for items with a small number of images or a single image and no associated OCR files (etc...) the separate bundle is a problem for the reasons you suggest, since it creates a hurdle and serves no necessary propose. I think this issue came up in the slack chat. 

Per the Search API, I agree that this is complicated and shouldn't be required to use DSpace IIIF. More complete DSpace support for the Search API seems possible but I agree that needs to come later.  In this PR the IIIFSearchable `entity.type` adds the search service in the manifest. The plain IIIF `entity.type` does not. We can easily find another approach to making the switch as well as provide a default setting. There is a similar switch taking place in the Angular component, so the user is only presented with a search option when the item metadata (`entity.type` for now) indicates that search is needed. 

So I guess overall I want to be sure we find an effective strategy to manage large documents (that will typically be imported via AIP as the most convenient approach).  And for now I think it's fine if the Search API is only an option that can be chosen by the adventurous. That does mean that a configurable search endpoint is included in our current API. 

We haven't discussed the IIIF image server. My experience is that setting up the Cantaloupe image server (my choice) is quite simple and can be described with a few brief instructions. But it is the responsibility of the user and depends a bit on the user's technical skills. It's similar to enabling solr for DSpace.

mspalti commented:
Hi all, I had time today to think about this more.

I definitely agree with Andrea's and the 4Science recommendation to eliminate the bundle constraint for the general use case and I think it scales well to many images when the data is homogenous. The problem I see is when an institution wants to add digitalized documents to DSpace and expose them to the world using IIIF. This is a common scenario, and the problem is that if we use a single bundle things get messy. There will often be images, metadata files of various types, and a PDF version of the entire document. And as I mentioned, there can be hundreds of these files.  Many will want to add these files to the record for management and preservation, and I think many will also want so share the data via IIIF using image canvases, annotation lists, and renderings. I'm trying to see how we could do all of this relying only on DSO metadata. It is technically possible but seems humanly daunting.

So my thought is that multiple bundles are useful for more complex IIIF scenarios. Would it make sense to you all if, in the more general use case, all data resides in the main bundle and is identified with something like `dspace.iiif:simple`? This would support a DSpace collection with existing image records, those who want to add image records via the GUI, and similar things that should be easy to do. 

The more complex use case would distribute bitstreams into additional bundles, identified using something like `dspace.iiif:bundled` This would require that content be added using an ingest method, which, frankly, is probably what an advanced user with complex data will want anyway.

In terms of what this means for implementation, the Mirador viewer would be embedded in the record when either `dspace.iiif:simple` or `dspace.iiif:bundled` is present.  On the REST side, the  values tell the service either to look for everything in the ORIGINAL bundle or to look to the separate bundles.

mspalti committed:
Modified search service.

mspalti committed:
Mostly code improvements with one functional change in how the iiif bundle is chosen.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Code cleanup with comments and missing license header.

mspalti committed:
Updated IT.

mspalti committed:
Minor comment update.

mspalti commented:
I spent some time yesterday and today improving the code. Not super relevant here since I assume much will be replaced by 4Science. But here are a few observations that came from the exercise.

- A simple change made it possible to default to the ORIGINAL bundle when a separate IIIF bundle is not provided. It's now possible to create an IIIF item using the GUI, etc. 
- Using the ORIGINAL bundle as the fallback probably means we can avoid the `dspace.iiif:bundled` that I was pondering in my previous comment.
- The `dspace.entity.type` still triggers the Mirador viewer in the Angular item view.  And the REST API still expects it to be in the metadata. I don't want to remove that constraint but we need another metadata element as the condition.  I'm sure the 4Science folks have considered other options for enabling IIIF and, yes, there might be a better approach.
- The seeAlso Annotation (as in this link https://digitalcollections.willamette.edu/server/iiif/413065ef-e242-4d0e-867d-8e2f6486be56/manifest/seeAlso ) still expects files in the `OtherContent` bundle. There's no option to fall back to ORIGINAL and I'm not sure that there should be. But this would be possible.
- The `IIIF` and `OtherContent` bundle names are hard-coded. These could be defined in configuration instead and overriden if local circumstances dictate. I'm still convinced that for more complex items the two swim lanes are going to be necessary.
-  The search option is currently triggered in both the Mirador viewer and in the IIIF manifest when the `dspace.entity.type` is set to IIIFSearchable. To support a search option we'll need a metadata field in our dspace schema to replace that entity. 
- Search as I've defined it here is based on an external solr plugin and there's no (current) dspace support for indexing. So it's for the adventurous and for experimentation and development. Since search must be explicitly enabled for the item I don't think it will cause confusion. It can be described as "experimental" in documentation.

mspalti committed:
Service boundary and model updates.

mspalti committed:
Updated range and image services.

mspalti committed:
Fixed IT.

mspalti committed:
Reverted manifest thumbnail image.

mspalti committed:
Fixed runaway sequence.

mspalti committed:
Javadoc updates.

mspalti committed:
Minor fix in sequence and an implementation comment in the manifest service.

mspalti committed:
Removed unused import.

mspalti commented:
I'm mostly on vacation next week. I'm happy to respond to discussion or code review but won't be making code changes until I get back. From my perspective this is in pretty good shape now.  

Aside from making the IIIF bundle optional, other changes we're discussing await the 4Science team's contribution. @abollini let me know if there's anything else I can do in the meantime. fyi, I modified code in a local branch to work with DSO metadata. It looks like this will be easy to do using my current implementation but it's not something we can actually test right now without other work.

abollini commented:
We have gone deeply in the code and it looks very good. I need to highlight that it uses a couple of new concepts from Spring that are not trivial and not usual for DSpace.
The source code uses request-scoped and prototype beans and the spring caching layer. For the request-scoped beans it would be probably sufficient add some extra javadoc/comments to warn the developers about their special behavior when injected in singleton bean (a proxy is used by spring so that each method call to the request scoped injected bean goes indeed to the right-request instance) . The use of a caching layer is interesting but there are aspects not yet implemented properly such as the cache invalidation when an item is modified or the cache size control. If we will be not able to implement such aspects I would recommend to remove the caching layer in the first version.

From the functional perspective I like to anticipate the changes that we are going to introduce to double check that they are liked by @mspalti first but also by others.

1. we want to switch from item to bundle as entry point for a IIIF manifest. In the basic scenario an item will have just a single ORIGINAL bundle but doing this would support more complex use case where a single item is used to describe for instance multivolume books that would be better represented by separate IIIF Manifests. The metadata in the IIIF Manifest will be populated with both the bundle metadata and the item metadata. Setting custom metadata at the bundle level would require extra effort and technical skills but it is technical feasible, basic users will just use the item metadata. A simple endpoint to retrieve all the available manifest for an item can be implemented so that the angular UI will just use it to present the viewer
2. metadata mapping will be delegated to an injectable service with a default implementation based on an external configuration file, so that the repository manager can handle exactly which information are exposed in the manifest
3. structure / toc in the manifest will be based on the metadata stored in the individual bitstream (iiif.bitstream.toc), also in this case we want to inject a TocService. The metadata approach will be used by the default implementation
4. default for view hint, height and width will be moved to technical metadata (iiif.default.viewhint, etc.) so that configuration values can be modified on individual items or bundles
5. manifest will include only anonymous readable images and content
6. additional rendering / related content is expected to be found in the same bundle where the images are placed. Bitstreams will be processed guessing their purpose (image, rendering, related content) from the mimetype or a technical metadata (iiif.bitstream.type) if present

Finally we are going to refactor the `IIIFRestRepository` in a IIIFService as it is indeed not a (DSpaceRest)Repository. To avoid code duplication the `obtainContext` method will be moved to the `ContextUtil` class

mspalti committed:
Made the search service a configurable plugin.

mspalti committed:
Removed remaining use of generator bean factories.

mspalti commented:
@abollini, here are my thoughts on the functional changes. They look good to me overall!

Using the IIIF bundle as the starting point for the manifest is interesting. I admit that I hadn't thought of the multivolume book scenario and I can see that something like this approach is needed to support that situation. If I understand correctly, most of the time there will be an item with an ORIGINAL bundle, but your implementation accommodates more complicated publications as needed. I think that makes good sense.

Points 2, 3, 4 all sound good.

Point 5 is important. I obviously agree, but I'd like to discuss a special case. I know of situations where the user wants the image to be displayed in the viewer while restricting download of high-resolution images. I can imagine a way to do this by removing anonymous bitstream access and granting read access to a special group assigned to requests that come from the IIIF image server IP. That way the high-res image could not be downloaded by anonymous users but it could be viewed by anyone via the IIIF viewer. As implemented on the Angular side the Mirador viewer includes a download plugin that could be disabled when the bitstream is not anonymous read. Does something like this sound feasible to you? 

Point 6 is one we've been discussing. I agree that the single bundle approach will work and I can see it's probably required in the multivolume book scenario. I'm still concerned about the potentially large number of bitstreams in a single bundle and its impact on usability, since I imagine this would make more it difficult to do things like update `iiif.bitstream.label` or `iiif.bitstream.toc`. Also, the correct order of image bitstreams is essential. If 4Science has found that this isn't a usability concern in your work with clients I'm ok with adding everything to the same bundle.

mspalti committed:
Fixed the canvas identifier in search annotions.

mspalti committed:
Moved default iiif configuration to iiif.cfg.

mspalti committed:
Correction to iiif cors configuration.

abollini commented:
@mspalti I'm happy to hear that want we propose seems ok so we will proceed to make the actual changes and create a PR for you in the next days.
About your last point, yes having ton of bitstreams not really matter if in multiple bundle or in a single one is not trivial for usability as it is often unpractical to deal with them on an UI. Multiple bundle sometime can help reducing the size if instead than 1000 images to manage at once you can have 10 set of 100 maybe it can be supported. In our addon we have implemented extra functionalities for batch managing that help on these scenarios, these can be considered as improvements for future versions. Having the information in bitstream metadata in any case provide immediate support for less problematic use cases and allow advanced institution to support batch updating / reordering via some scripts.
The position of the additional rendering bitstreams doesn't really matter, they can be at the start, at the end or also mixed with the images as the manifest generator will need in any case to scan all the bitstreams.

abollini committed:
Refactor the IIIFRestRepository to avoid inheritance from DSpaceRestRepository

mspalti committed:
Updated javadoc.

abollini committed:
Intermediate commit - refactoring to rely on metadata for feature activation and configuration

abollini committed:
Fix range generation and improve basic structure test

mspalti committed:
Updated javadocs and changed generator interface.

mspalti committed:
Additional documentation in iiif.cfg file.

mspalti committed:
Updated experimentatal search plugin support to use full package name.

abollini committed:
Finalize configuration via metadata refactoring and hierarchical range support

abollini commented:
For anyone else that is following this PR be aware that me and @mspalti are working on some refactoring in a side PR https://github.com/mspalti/DSpace/pull/1

mspalti committed:
Added not implemented exception for search service.

mspalti committed:
Search service update.

mspalti committed:
Updated imports for style check.

mspalti committed:
Minor iiif.cfg updates.

mspalti committed:
Merge branch '4Science-pr' into iiif-endpoint

mspalti committed:
More updates after merging PR#1.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Fixed lgtm alerts.

mspalti commented:
@abollini , I reviewed, made a few minor changes, and merged PR #1. 

All of the ITs for IIIF support are passing but there seems to be an error with localization in SubmissionFormsControllerIT. You may need to take a look at this one.

abollini committed:
fix locale setting in the obtainContext method

abollini committed:
Add test to proof behavior with restricted items

mspalti committed:
Added null check in canvas service to prevent exception in iiif metadata model.

mspalti committed:
Added todo notice to cache configuration class.

abollini committed:
Merge branch 'iiif-endpoint' of https://github.com/mspalti/DSpace into iiif-endpoint

mspalti committed:
Fixed identifiers for range canvas references.

mspalti committed:
Excluding OtherContent bundle from bundles that can contain iiif image resources.

tdonohue commented on dspace/config/modules/iiif.cfg:
```diff
@@ -0,0 +1,93 @@
+# these should be not changed. They must match the routing configuration for the iiif controller 
+iiif.url = ${dspace.server.url}/iiif
+iiif.solr.search.url =
+iiif.bitstream.url = ${dspace.server.url}/api/core/bitstreams
```
Silly question: If these configurations should not be changed, then why are they configurations?  Shouldn't they be constants in the code somewhere?  Also unclear why `iiif.solr.search.url` is set to empty.

mspalti replied:
I'll look into this. It seems that they should be in code as you suggest.

tdonohue commented on dspace/config/registries/iiif-types.xml:
```diff
@@ -0,0 +1,66 @@
+        <schema>iiif</schema>
+        <element>search</element>
+        <qualifier>enabled</qualifier>
+        <scope_note>Metadata field used to enable the IIIF Search service at the item level</scope_note>
```
Description needs updating, as this doesn't seem to be item specific

mspalti replied:
This is used to configure items to have the search service added to the manifest. Maybe: "Metadata field used  at the item level to add the IIIF Search service annotation to the manifest for that item."

tdonohue replied:
Oh, whoops, I just realized my initial comment was wrong. I mixed up this `iiif.search.enabled` field with the `dspace.iiif.enabled` field...the latter is the one which is not Item specific.  But, I just rereviewed both the fields' scope_note settings, and they both are fine as-is. Sorry for the confusion.  I'll flag this conversation as resolved.

tdonohue commented on dspace/config/modules/solr-statistics.cfg:
```diff
@@ -13,7 +13,7 @@
 
 # A comma-separated list that contains the bundles for which the bitstreams will be displayed
-solr-statistics.query.filter.bundles=ORIGINAL
+solr-statistics.query.filter.bundles=ORIGINAL, IIIF
```
Do we have an easy way to add a IIIF bundle? Or is that just expected to occur by editing an Item and uploading to a bundle of that name?  Just trying to understand the process of using this bundle and whether it will be "connected" into the submission process eventually.

mspalti replied:
The IIIF bundle needs to be removed from this configuration.   

The 4Science approach no longer uses a designated bundle for IIIF. I think the current submission process and UI will accommodate the typical case quite well, especially after adding of a few new metadata fields for bitstreams.  That addition hasn't been proposed yet but it might be something to add in 7.1.

In the new approach there can (optionally) be multiple bundles and bundles can carry IIIF-related metadata. @abollini knows more about this since it is part of the 4Science implementation.

tdonohue reviewed (changes requested):
@mspalti and @abollini :  Gave this a first code review today. I have not tested it yet, but rather just reviewed it from a code best practices standpoint.  While I have a lot of inline comments, some of these are simple questions, as something in the code didn't make sense to me....and it's very possible some of these questions may come from my lack of detailed knowledge on IIIF (I know basic concepts, but haven't used it in any production scenario). So, apologies if any of the questions are basic misunderstandings :) 

Overall, the code is good quality and looks well structured (though I think some of the packages could be renamed, see below).  Thanks again for all your hard work on this!

mspalti committed:
Cache add.

mspalti committed:
Completed consumer used to manage iiif cache.

mspalti committed:
Added comment to dspace.cfg.

mspalti commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/Application.java:
```diff
@@ -157,16 +157,34 @@ public WebMvcConfigurer webMvcConfigurer() {
+                String[] corsAllowedOrigins = configuration
+                    .getCorsAllowedOrigins(configuration.getCorsAllowedOriginsConfig());
+                String[] iiifAllowedOrigins = configuration
+                    .getCorsAllowedOrigins(configuration.getIiifAllowedOriginsConfig());
```
The "interoperability" in IIIF requires us to be able to provide resources to domains outside of DSpace. For example, if you are incorporating images from DSpace (along with images from other repositories) into another IIIF project. Separate CORS configuration can accommodate this.  

I agree that the paths are worth considering.

mspalti commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/IIIFController.java:
```diff
@@ -0,0 +1,113 @@
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest;
```
I don't see a reason not to do this in general; it actually sounds better to me.

abollini replied:
I agree as well, thanks to point this out

mspalti commented on dspace-server-webapp/src/main/java/org/dspace/app/rest/iiif/IIIFServiceFacade.java:
```diff
@@ -0,0 +1,133 @@
+ *
+ * http://www.dspace.org/license/
+ */
+package org.dspace.app.rest.iiif;
```
ditto

mspalti commented on dspace-server-webapp/src/main/resources/application.properties:
```diff
@@ -89,6 +89,12 @@ server.error.include-stacktrace = never
+# Cache Properties
+# Added for IIIF cache support.
+# Path to configuration file.
+spring.cache.jcache.config=classpath:iiif/cache/ehcache.xml
```
At the suggestion of @abollini I'm just about to commit a Consumer for review. It evicts things from the manifest cache when an Item, Bundle or Bitstream is modified. So far now, the cache stays...

mspalti commented on dspace/config/modules/iiif.cfg:
```diff
@@ -0,0 +1,93 @@
+# Base URL of the search service used to support the optional IIIF Search
+# capability. The actual path will depend on how search is supported.
+# (experimental)
+# iiif.search.url = ${solr.server}/solr/word_highlighting
```
You are right. This needs to be changed.

mspalti committed:
Updated consumer.

mspalti committed:
Configuration update.

mspalti commented:
I implemented the iiif event consumer for evicting from the "manifests" cache.  It works fine when using the the web application.

Unfortunately, because the cache is managed by the spring web application, command line dspace-api tasks throw a `ClassNotFoundException` exception when creating the iiif consumer profile.  @tdonohue or @abollini , if you have ideas on how to handle this I'd appreciate your insight. At the moment, the only solution is to disable temporarily the iiif consumer in `dspace.cfg`.

In general, I think the iiif consumer is not necessary when doing batch work from the command line. I like the idea of a protected rest endpoint that the admin could use to clear the cache, if they need to, after a batch operation.

abollini commented:
@mspalti if you move all the dependencies of the `IIIFCacheEventConsumer` into the dspace-api project you should be able to get ride of the class not found exception.
The methods in `CacheEvictService` needs to be protected against NPE as the cache will be present only when processing request to the web application but not when the consumer will be hit by a batch process. A method `isCacheEnabled` in the service could inform the consumer so that all the logic is skipped without unnecessary overhead.
We should also add a test specific for the caching mechanism, once that you enable the consumer by default you should be able to demonstrate that the manifest of an item reflect changes to its metadata/bitstreams. The basic sequence should be
1. create the item
2. check the manifest
3. update the item
4. verify that the manifest reflect the changes
point 4 should be broken without the consumer.

Let me know if we should jump in on that

mspalti committed:
Moved iiif event consumer to dspace-api.

mspalti commented:
Thanks @cwilper .  I'll update the site to use the latest versions and provide new URLs later today.

mspalti committed:
Updated event consumer.

mspalti committed:
Added "org.dspace.iiif" to component scan in ApplicationConfig.

mspalti committed:
Added component annotations.

Added comment.

Fixed import.

mspalti committed:
Added iiif to default consumers.

mspalti committed:
Work on test for iiif cache consumer.

mspalti committed:
Updated iiif configuation.

mspalti committed:
Javadoc for iiif interfaces.

mspalti committed:
Moved iiif to separate module.

mspalti committed:
Using dspace-api ContextUtil with iiif controller.

mspalti committed:
Docs update.

mspalti committed:
Added iiif consumer to default dispatcher.

mspalti committed:
Adding rendering to manifest instead of sequence.

abollini committed:
Fix consumer to avoid early initialization of the iiif cache evict service

abollini committed:
Add attributions to the main iiif classes

abollini commented:
@tdonohue @mspalti I have fixed the iiif consumer and re-enabled the cache test. It should be ready to go now

mspalti committed:
Updated to a method signature.

tdonohue reviewed (commented):
@mspalti : Tested this with the frontend & it works great. Still reviewing all the new code here though... looks great overall, but I wanted to submit a few early comments of things I've found. 

I **do believe** we need a way to disable this feature.  It shouldn't be enabled by default, as it requires Cantaloupe or similar to be installed.  See below.

tdonohue reviewed (commented):
@mspalti (and @abollini) : Code looks good, just a few more minor notes (see below) to add to [my previous comments from today](https://github.com/DSpace/DSpace/pull/3210#pullrequestreview-782616155).  

👍 Overall, I **approve** of these changes, but I think that we need to find a way to disable IIIF by default.  I gave some suggestions in my prior feedback around how to do so.  If there's an issue with this, let me know though...I'm hoping it won't be a big change though.

mspalti committed:
Optional iiif configuration and javadoc.

mspalti committed:
Unused import in test.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Updated classes to call ContextUtil.obtainContext with HttpServletRequest objects.

mspalti commented:
In this PR @abollini updated `ContextUtil.obtainContext` to take `HttpServletRequest` as the argument (rather than `ServletRequest`). It looks like 3 recently merged classes need to be updated to work with that change. I'm running the local tests now and will push after they succeed.

mspalti committed:
Unused import.

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Merge branch 'main' into iiif-endpoint

mspalti committed:
Removed unused property.

mspalti committed:
Minor comment change.

tdonohue reviewed (approved):
👍 Thanks @mspalti and @abollini for all the hard work here!  This all looks great to me now.  I've tested that the `/iiif/[uuid]/manifest` URLs don't respond by default.  But, that everything still works great once you set `iiif.enabled = true` and add the `iiif` consumer to `event.dispatcher.default.consumers`.

This looks great to me & I agree this is ready to merge.

abollini reviewed (approved):
The PR is in a good state for me. There is of course room for improvements but these can wait for a 7.1
